check-by-name: Refactor to prepare for enforcing `pkgs/by-name`, make `--base` required

infinisil opened the pull request:
## Description of changes

This is a preparation PR for https://github.com/NixOS/nixpkgs/pull/275539, including only internal improvements to the code and comments, without changing any functionality. This makes the diff for the above PR much smaller.

The only somewhat interesting change is that `--base` is now required for `nixpkgs-check-by-name`, which was already prepared for in https://github.com/NixOS/nixpkgs/pull/274591 by always passing the flag in CI.

This PR is best reviewed commit-by-commit.

## Things done

- [x] Built and tested on `x86_64-linux`

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

infinisil committed:
tests.nixpkgs-check-by-name: Refactor eval code and improve comments

Does a bunch of cleanups to the eval.{rs,nix} code to make future
changes easier, no functionality is changed.

infinisil committed:
tests.nixpkgs-check-by-name: Set evaluation system to x86_64-linux

This was previously a checking impurity that could produce different
results when run on different systems.

infinisil committed:
tests.nixpkgs-check-by-name: Make --base required

CI now passes the flag, so it doesn't have to be optional anymore

infinisil committed:
tests.nixpkgs-check-by-name: Re-usable ratchet logic

This makes the attribute ratchet check logic more re-usable, which will
be used in a future commit.

It also renames the ratchet states to something more intuitive

infinisil committed:
tests.nixpkgs-check-by-name: Internal strip nixpkgs prefix

Strips the Nixpkgs prefix from the callPackage paths,
makes future error messages using this path be deterministic.

infinisil committed:
tests.nixpkgs-check-by-name: Deterministic ordering

Makes errors for attributes deterministic so it's easier to test (also,
reproducibility is always nice)

philiptaron commented on pkgs/test/nixpkgs-check-by-name/tests/empty-base/pkgs/by-name/README.md:
I'm assuming this is a sentinel file.

infinisil replied:
Yeah, just so that git can track that the directory exists. Marginally easier than changing the code to handle a missing directory

philiptaron commented on pkgs/test/nixpkgs-check-by-name/src/eval.rs:
```diff
@@ -6,33 +6,48 @@ use std::path::Path;
 #[derive(Deserialize)]
-struct AttributeInfo {
-    variant: AttributeVariant,
+enum ByNameAttribute {
+    /// The attribute doesn't exist at all
+    Missing,
+    Existing(AttributeInfo),
+}
```
These types are much clearer than what came before.

philiptaron commented on pkgs/test/nixpkgs-check-by-name/src/eval.rs:
```diff
@@ -103,74 +118,88 @@ pub fn check_values(
+            use CallPackageVariant::*;
+
+            let check_result = match attribute_value {
+                Missing => NixpkgsProblem::UndefinedAttr {
+                    relative_package_file: relative_package_file.clone(),
+                    package_name: attribute_name.clone(),
+                }
+                .into(),
```
This looks like it's implementing `IntoNixpkgsProblem` on `ByNameAttribute`.

The same thing holds for the cases below on their specific types.

infinisil replied:
Not entirely because in some cases there won't be an error.

philiptaron replied:
> Not entirely because in some cases there won't be an error.

Agreed. I took a swing at this (PR not sent out) and ran into the same thing. There's some fine tuning that could be done for sure but I think the trait-based approach I suggested isn't right.

philiptaron commented on pkgs/test/nixpkgs-check-by-name/src/eval.nix:
I only lightly reviewed this file. I didn't see anything that caused me to believe that it's causing the tests to be misleading, and appears to be monkey-patching or mocking callPackage well enough.

philiptaron commented on pkgs/test/nixpkgs-check-by-name/src/eval.rs:
```diff
@@ -103,74 +118,88 @@ pub fn check_values(
     // Parse the resulting JSON value
-    let actual_files: HashMap<String, AttributeInfo> = serde_json::from_slice(&result.stdout)
+    let attributes: Vec<(String, ByNameAttribute)> = serde_json::from_slice(&result.stdout)
         .context(format!(
             "Failed to deserialise {}",
             String::from_utf8_lossy(&result.stdout)
         ))?;
```
Here and throughout the crate: use `anyhow::Context::with_context` instead of `context` to avoid allocating these error strings in the non-error case.

infinisil replied:
Done throughout the codebase in https://github.com/NixOS/nixpkgs/pull/278805/commits/a1db0cdf9bb5e90727eb1c455da448fb21bab91e :)

philiptaron reviewed (approved):
OK, I took a look at the Rust code here, and skimmed the test-only Nix code.

I have three non-blocking refactoring suggestions, in descending order of "want to see":

1. Simplify the traits in `ratchet.rs` and use them more broadly.
2. Use `anyhow::Context::with_context` to avoid allocating error strings in the non-error case.
3. Minor typo and comment fixes as noted.

There's a slightly larger refactoring that I see that's not directly related to the changes in this PR. In main, everywhere the `error_writer` is used could instead use a `ratchet::Nixpkgs` that's cased on its trivial success.

Ericson2314 commented on pkgs/test/nixpkgs-check-by-name/src/eval.nix:
```diff
@@ -1,82 +1,86 @@
-        variantInfo._attributeVariant = {
-          # These names are used by the deserializer on the Rust side
-          CallPackage.path =
+      addVariantInfo (prev.callPackage fn args) {
```
uhhh how is this in scope?

Ericson2314 reviewed (approved):
LGTM!

infinisil commented on pkgs/test/nixpkgs-check-by-name/src/ratchet.rs:
```diff
@@ -57,29 +46,61 @@ impl Package {
+
+/// A trait for a specific ratchet check for an attribute.
+trait AttributeRatchet: Sized {
+    /// What error to produce in case the ratchet went in the wrong direction
+    fn to_error(name: &str, context: &Self, existed_before: bool) -> NixpkgsProblem;
```
It looks like making it consume `Self` is a bit trickier, since the values come from a HashMap originally. And not sure what you mean with `context` being `self`!

infinisil committed:
tests.nixpkgs-check-by-name: Minor improvements from feedback

- Typo
- Rename AttributeRatchet to ToNixpkgsProblem
- Make the compare trait method into a RatchetState method

Co-Authored-By: Philip Taron <philip.taron@gmail.com>

infinisil committed:
tests.nixpkgs-check-by-name: .context -> .with_context

Avoids allocation in the non-error case

infinisil commented:
Feedback addressed, I'll merge once CI passes since the feedback isn't blocking :)

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/2024-01-09-nixpkgs-architecture-team-meeting-47/38037/1

philiptaron reviewed (approved):
Ship it.
